File: fibre/src/apis/dummy/gangster.go
package dummy

import (
	"database/sql"
	"fmt"
	"log"
	"github.com/gofiber/fiber/v2"
	 _"github.com/mattn/go-sqlite3"
)

type Gangster struct {
	GangsterID   string `json:"gangster_id"`
	GangsterName string `json:"gangster_name"`
}

type Response struct {
	Success bool        `json:"success"`
	Message string      `json:"message"`
	Data    interface{} `json:"data,omitempty"`
}

var db *sql.DB

// InitDB initializes the database connection
func InitDB() error {
	var err error
	db, err = sql.Open("sqlite3", "./gangsters.db")
	if err != nil {
		return err
	}

	// Create table if not exists
	createTableSQL := `CREATE TABLE IF NOT EXISTS gangsters (
		gangster_id TEXT PRIMARY KEY,
		gangster_name TEXT NOT NULL UNIQUE
	);`

	_, err = db.Exec(createTableSQL)
	if err != nil {
		return err
	}

	log.Println("✅ Database initialized successfully")
	return nil
}

// generateGangsterID generates a unique gangster ID like G01, G02, etc.
func generateGangsterID() (string, error) {
	var count int
	err := db.QueryRow("SELECT COUNT(*) FROM gangsters").Scan(&count)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("G%02d", count+1), nil
}

// GangsterBlueprint - Blueprint for gangster routes
func GangsterBlueprint(router fiber.Router) {

	// POST - Add new gangster
	router.Post("/add", func(c *fiber.Ctx) error {
		gangsterName := c.Get("gangster")

		if gangsterName == "" {
			return c.Status(fiber.StatusBadRequest).JSON(Response{
				Success: false,
				Message: "gangster header is required",
			})
		}

		// Generate gangster ID
		gangsterID, err := generateGangsterID()
		if err != nil {
			return c.Status(fiber.StatusInternalServerError).JSON(Response{
				Success: false,
				Message: "Failed to generate gangster ID",
			})
		}

		// Insert into database
		_, err = db.Exec("INSERT INTO gangsters (gangster_id, gangster_name) VALUES (?, ?)", gangsterID, gangsterName)
		if err != nil {
			return c.Status(fiber.StatusConflict).JSON(Response{
				Success: false,
				Message: "Gangster already exists or database error",
			})
		}

		return c.Status(fiber.StatusCreated).JSON(Response{
			Success: true,
			Message: "Gangster added successfully",
			Data: Gangster{
				GangsterID:   gangsterID,
				GangsterName: gangsterName,
			},
		})
	})

	// GET - Get all gangsters
	router.Get("/gangsters", func(c *fiber.Ctx) error {
		rows, err := db.Query("SELECT gangster_id, gangster_name FROM gangsters")
		if err != nil {
			return c.Status(fiber.StatusInternalServerError).JSON(Response{
				Success: false,
				Message: "Failed to fetch gangsters",
			})
		}
		defer rows.Close()

		var gangsters []Gangster
		for rows.Next() {
			var g Gangster
			if err := rows.Scan(&g.GangsterID, &g.GangsterName); err != nil {
				continue
			}
			gangsters = append(gangsters, g)
		}

		return c.JSON(Response{
			Success: true,
			Message: "Gangsters fetched successfully",
			Data:    gangsters,
		})
	})

	// GET - Get single gangster by ID
	router.Get("/gangsters/:id", func(c *fiber.Ctx) error {
		id := c.Params("id")

		var g Gangster
		err := db.QueryRow("SELECT gangster_id, gangster_name FROM gangsters WHERE gangster_id = ?", id).Scan(&g.GangsterID, &g.GangsterName)
		if err == sql.ErrNoRows {
			return c.Status(fiber.StatusNotFound).JSON(Response{
				Success: false,
				Message: "Gangster not found",
			})
		} else if err != nil {
			return c.Status(fiber.StatusInternalServerError).JSON(Response{
				Success: false,
				Message: "Database error",
			})
		}

		return c.JSON(Response{
			Success: true,
			Message: "Gangster fetched successfully",
			Data:    g,
		})
	})

	// PUT - Update gangster
	router.Put("/gangsters/:id", func(c *fiber.Ctx) error {
		id := c.Params("id")
		newName := c.Get("gangster")

		if newName == "" {
			return c.Status(fiber.StatusBadRequest).JSON(Response{
				Success: false,
				Message: "gangster header is required",
			})
		}

		result, err := db.Exec("UPDATE gangsters SET gangster_name = ? WHERE gangster_id = ?", newName, id)
		if err != nil {
			return c.Status(fiber.StatusInternalServerError).JSON(Response{
				Success: false,
				Message: "Failed to update gangster",
			})
		}

		rowsAffected, _ := result.RowsAffected()
		if rowsAffected == 0 {
			return c.Status(fiber.StatusNotFound).JSON(Response{
				Success: false,
				Message: "Gangster not found",
			})
		}

		return c.JSON(Response{
			Success: true,
			Message: "Gangster updated successfully",
			Data: Gangster{
				GangsterID:   id,
				GangsterName: newName,
			},
		})
	})

	// DELETE - Delete gangster
	router.Delete("/gangsters/:id", func(c *fiber.Ctx) error {
		id := c.Params("id")

		result, err := db.Exec("DELETE FROM gangsters WHERE gangster_id = ?", id)
		if err != nil {
			return c.Status(fiber.StatusInternalServerError).JSON(Response{
				Success: false,
				Message: "Failed to delete gangster",
			})
		}

		rowsAffected, _ := result.RowsAffected()
		if rowsAffected == 0 {
			return c.Status(fiber.StatusNotFound).JSON(Response{
				Success: false,
				Message: "Gangster not found",
			})
		}

		return c.JSON(Response{
			Success: true,
			Message: "Gangster deleted successfully",
		})
	})
}

File: actix/src/apis/dummy/gangster.rs
use actix_web::{web, HttpRequest, HttpResponse, Responder};
use rusqlite::{Connection, Result};
use serde::{Deserialize, Serialize};
use std::sync::Mutex;

#[derive(Serialize, Deserialize, Clone)]
pub struct Gangster {
    gangster_id: String,
    gangster_name: String,
}

#[derive(Serialize)]
struct Response<T> {
    success: bool,
    message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    data: Option<T>,
}

pub fn init_db() -> Result<Connection> {
    let conn = Connection::open("gangsters.db")?;

    conn.execute(
        "CREATE TABLE IF NOT EXISTS gangsters (
            gangster_id TEXT PRIMARY KEY,
            gangster_name TEXT NOT NULL UNIQUE
        )",
        [],
    )?;

    Ok(conn)
}

fn generate_gangster_id(conn: &Connection) -> Result<String> {
    let count: i32 = conn.query_row("SELECT COUNT(*) FROM gangsters", [], |row| row.get(0))?;
    Ok(format!("G{:02}", count + 1))
}

// POST - Add new gangster
async fn add_gangster(
    req: HttpRequest,
    db: web::Data<Mutex<Connection>>,
) -> impl Responder {
    let gangster_name = match req.headers().get("gangster") {
        Some(val) => match val.to_str() {
            Ok(v) => v.to_string(),
            Err(_) => {
                return HttpResponse::BadRequest().json(Response::<()> {
                    success: false,
                    message: "Invalid gangster header".to_string(),
                    data: None,
                })
            }
        },
        None => {
            return HttpResponse::BadRequest().json(Response::<()> {
                success: false,
                message: "gangster header is required".to_string(),
                data: None,
            })
        }
    };

    let conn = db.lock().unwrap();

    let gangster_id = match generate_gangster_id(&conn) {
        Ok(id) => id,
        Err(_) => {
            return HttpResponse::InternalServerError().json(Response::<()> {
                success: false,
                message: "Failed to generate gangster ID".to_string(),
                data: None,
            })
        }
    };

    match conn.execute(
        "INSERT INTO gangsters (gangster_id, gangster_name) VALUES (?1, ?2)",
        [&gangster_id, &gangster_name],
    ) {
        Ok(_) => HttpResponse::Created().json(Response {
            success: true,
            message: "Gangster added successfully".to_string(),
            data: Some(Gangster {
                gangster_id,
                gangster_name,
            }),
        }),
        Err(_) => HttpResponse::Conflict().json(Response::<()> {
            success: false,
            message: "Gangster already exists or database error".to_string(),
            data: None,
        }),
    }
}

// GET - Get all gangsters
async fn get_gangsters(db: web::Data<Mutex<Connection>>) -> impl Responder {
    let conn = db.lock().unwrap();

    let mut stmt = match conn.prepare("SELECT gangster_id, gangster_name FROM gangsters") {
        Ok(s) => s,
        Err(_) => {
            return HttpResponse::InternalServerError().json(Response::<()> {
                success: false,
                message: "Failed to fetch gangsters".to_string(),
                data: None,
            })
        }
    };

    let gangsters_iter = stmt.query_map([], |row| {
        Ok(Gangster {
            gangster_id: row.get(0)?,
            gangster_name: row.get(1)?,
        })
    });

    let gangsters: Vec<Gangster> = match gangsters_iter {
        Ok(iter) => iter.filter_map(Result::ok).collect(),
        Err(_) => {
            return HttpResponse::InternalServerError().json(Response::<()> {
                success: false,
                message: "Failed to fetch gangsters".to_string(),
                data: None,
            })
        }
    };

    HttpResponse::Ok().json(Response {
        success: true,
        message: "Gangsters fetched successfully".to_string(),
        data: Some(gangsters),
    })
}

// GET - Get single gangster by ID
async fn get_gangster(
    path: web::Path<String>,
    db: web::Data<Mutex<Connection>>,
) -> impl Responder {
    let id = path.into_inner();
    let conn = db.lock().unwrap();

    match conn.query_row(
        "SELECT gangster_id, gangster_name FROM gangsters WHERE gangster_id = ?1",
        [&id],
        |row| {
            Ok(Gangster {
                gangster_id: row.get(0)?,
                gangster_name: row.get(1)?,
            })
        },
    ) {
        Ok(gangster) => HttpResponse::Ok().json(Response {
            success: true,
            message: "Gangster fetched successfully".to_string(),
            data: Some(gangster),
        }),
        Err(rusqlite::Error::QueryReturnedNoRows) => {
            HttpResponse::NotFound().json(Response::<()> {
                success: false,
                message: "Gangster not found".to_string(),
                data: None,
            })
        }
        Err(_) => HttpResponse::InternalServerError().json(Response::<()> {
            success: false,
            message: "Database error".to_string(),
            data: None,
        }),
    }
}

// PUT - Update gangster
async fn update_gangster(
    req: HttpRequest,
    path: web::Path<String>,
    db: web::Data<Mutex<Connection>>,
) -> impl Responder {
    let id = path.into_inner();

    let new_name = match req.headers().get("gangster") {
        Some(val) => match val.to_str() {
            Ok(v) => v.to_string(),
            Err(_) => {
                return HttpResponse::BadRequest().json(Response::<()> {
                    success: false,
                    message: "Invalid gangster header".to_string(),
                    data: None,
                })
            }
        },
        None => {
            return HttpResponse::BadRequest().json(Response::<()> {
                success: false,
                message: "gangster header is required".to_string(),
                data: None,
            })
        }
    };

    let conn = db.lock().unwrap();

    match conn.execute(
        "UPDATE gangsters SET gangster_name = ?1 WHERE gangster_id = ?2",
        [&new_name, &id],
    ) {
        Ok(rows) => {
            if rows == 0 {
                HttpResponse::NotFound().json(Response::<()> {
                    success: false,
                    message: "Gangster not found".to_string(),
                    data: None,
                })
            } else {
                HttpResponse::Ok().json(Response {
                    success: true,
                    message: "Gangster updated successfully".to_string(),
                    data: Some(Gangster {
                        gangster_id: id,
                        gangster_name: new_name,
                    }),
                })
            }
        }
        Err(_) => HttpResponse::InternalServerError().json(Response::<()> {
            success: false,
            message: "Failed to update gangster".to_string(),
            data: None,
        }),
    }
}

// DELETE - Delete gangster
async fn delete_gangster(
    path: web::Path<String>,
    db: web::Data<Mutex<Connection>>,
) -> impl Responder {
    let id = path.into_inner();
    let conn = db.lock().unwrap();

    match conn.execute("DELETE FROM gangsters WHERE gangster_id = ?1", [&id]) {
        Ok(rows) => {
            if rows == 0 {
                HttpResponse::NotFound().json(Response::<()> {
                    success: false,
                    message: "Gangster not found".to_string(),
                    data: None,
                })
            } else {
                HttpResponse::Ok().json(Response::<()> {
                    success: true,
                    message: "Gangster deleted successfully".to_string(),
                    data: None,
                })
            }
        }
        Err(_) => HttpResponse::InternalServerError().json(Response::<()> {
            success: false,
            message: "Failed to delete gangster".to_string(),
            data: None,
        }),
    }
}

pub fn gangster_blueprint(cfg: &mut web::ServiceConfig) {
    cfg.service(
        web::scope("")
            .route("/add", web::post().to(add_gangster))
            .route("/gangsters", web::get().to(get_gangsters))
            .route("/gangsters/{id}", web::get().to(get_gangster))
            .route("/gangsters/{id}", web::put().to(update_gangster))
            .route("/gangsters/{id}", web::delete().to(delete_gangster)),
    );
}
